src/Stratis.Bitcoin.Features.SignalR/Broadcasters/BroadcasterBase.cs

```diff
@@ -36,15 +39,38 @@ public void Init(ClientEventBroadcasterSettings broadcasterSettings)
                 $"Broadcast {this.GetType().Name}",
                 async token =>
                 {
-                    foreach (IClientEvent clientEvent in this.GetMessages())
+                    if (null == this.getMessagesTask)
                     {
-                        await this.eventsHub.SendToClientsAsync(clientEvent).ConfigureAwait(false);
+                        lock (this.lockObject)
+                        {
+                            if (null == this.getMessagesTask)
+                            {
+                                this.getMessagesTask = this.GetMessages();
```

Comment:
The issue with performance is not within BroadcasterBase but get GetMessages() method which has a synchronous sequence of invocations to wallet manager, consensus manager etc which with this code needs to complete and only after that 1st argument in this.getMessagesTask.ContinueWith() gets invoked. This would imply that if any invocation on the GetMessages() takes long time this all async loop gets delayed? Doesnt seem there is cancellation token in GetMessages(). Author should consider breaking down invocation of manager objects in GetMessages() into seperate async methods

Reply:
Would be good to get some more ideas on this. I don't see that there's a benefit to breaking down the calls in `WalletInfoBrodcaster.GetMessages()` and awaiting all the individual tasks as we are already on running on an async task here, so we are just adding cost with all the potentially new threads we are spawning an this in itself is expensive, it's going to look really messy and confusing also and raises a lot of questions like what if certain tasks fail etc.. 

The option I think you are suggesting is to have this broadcast object cached and then build up the signalR object piecemeal with the results from each of the tasks as and when they complete which is a lot of work I think and need to consider the priority of that especially as this now seems to be running fine since the SQLWallet Performance was fixed and we now only get Address Balances on the Cirrus Broadcaster. It was the slow performance of `address.GetBalances()` that high lighted this.

The fix I have provided will stop lots of tasks building up and saturating the task scheduler should we have a performance regression in the future.

As for the cancellation token we can add that, would need some clarity on the source of the cancellation token as the one provided by the AsyncLoop provider looks like it's at the Node level I guess for shutdown etc, is that the one you want using?

Reply:
to clarify I am suggesting to break them down, await them seperately and set some time limit per each invocation to complete(CancellationToken). That way you dont compromise sending out the whole set of data samples because one of the items didnt return. Your code assumes everything to operate efficiently which is not guaranteed therefore a chenge is required

Reply:
@ianadavies I agree with Lukasz here. We need to pass a cancellation token to the `ContinueWith` just to be on the safe side in case one of the get message tasks does not return. I would create a new cancellation token and set a timeout of about 10 seconds as well as link it to the node lifetime's cancellation token.

Reply:
Sorry the cancellation token should probably be on each `getMessages` task?